Setup.bat

```diff
@@ -27,14 +36,70 @@ call :MarkStartOfBlock "Setup the git hooks"
+
+    rem Find the Unreal Engine used to build this project.
+    set UNREAL_ENGINE=""
+
+    rem Get the Unreal Engine used by this project by querying the registry for the engine association found in the .uproject.
+    for /f "delims=" %%A in (' powershell -Command "Get-Childitem -Path %ProjectDirectory% -Recurse -Include *.uproject -File | %% {$_.FullName}" ') do set UPROJECT="%%A"
+    
+    rem If the regex failed then it will return a string containing only a space.
+    if %UPROJECT%==" " (
+        echo Error: Could not find uproject. Please make sure you have passed in the project directory correctly.
         pause
         exit /b 1
     )
 
+    echo Using uproject: %UPROJECT%
+
+    rem Get the Engine association from the uproject.
+    for /f "delims=" %%A in (' powershell -Command "Select-String -Pattern '..{8}-.{4}-.{4}-.{4}-.{12}.' %UPROJECT% -AllMatches | %% { $_.Matches } | %% { $_.Value }" ') do set ENGINE_ASSOCIATION="%%A"
+    
+    echo Engine association for uproject is: "%ENGINE_ASSOCIATION%"
+
+    if not "%ENGINE_ASSOCIATION%"=="" (
+        rem Query the registry for the path to the Unreal Engine using the engine associtation.
+        for /f "usebackq tokens=3*" %%A in (`reg query "HKCU\Software\Epic Games\Unreal Engine\Builds" /v %ENGINE_ASSOCIATION%`) do (
+            set UNREAL_ENGINE="%%A"
+        )
+    )
+
+    rem If there was no engine association then we need to climb the directory path of the project to find the Engine.
+    pushd "%~dp0"
+    cd /d  %ProjectDirectory%
+    if %UNREAL_ENGINE%=="" (
+        :climb_parent_directory
+        cd ..
+        if exist Engine (
+            rem Check for the Build.version file to be sure we have found a correct Engine folder.
+            if exist "Engine\Build\Build.version" (
+                echo Found Engine at: "%cd%"
+                set UNREAL_ENGINE="%cd%"
+            )
+        ) else (
+            rem This checks if we are in a root directory. If so we cannot check any higher and so should error out.
+            if "%cd:~3,1%"=="" (
+                echo Error: Could not find Unreal Engine folder. Please set a project association or ensure your game project is within an Unreal Engine folder.
+                pause
+                exit /b 1
+            )
+            goto :climb_parent_directory
+        )
+    )
+
+    popd
+    pushd "%~dp0"
+
+    if %UNREAL_ENGINE%=="" (
+        echo Error: Could not find the Unreal Engine. Please associate your '.uproject' with an engine version or ensure this game project is nested within an engine build.
+        pause
+        exit /b 1
+    )
+
+    echo Using Unreal Engine at: %UNREAL_ENGINE%
+
     rem Use Unreal Engine's script to get the path to MSBuild. This turns off echo so turn it back on for TeamCity.
-    call "%UNREAL_HOME%\Engine\Build\BatchFiles\GetMSBuildPath.bat"
+    call "%UNREAL_ENGINE%\Engine\Build\BatchFiles\GetMSBuildPath.bat"

```

Comment:
The quoting is inconsistent for `%UNREAL_ENGINE% - if the variable has quotes embedded in it, then this call would fail, right?

Reply:
Looked into this, it seems that double quotes are handled nicely in batch and so it's not an issue. It's possible to have quotes within a path (for paths that have spaces) and have that wrapped in another path that has quotes.